### PR TITLE
Introduce resource model pool to replace UniversalFactory

### PR DESCRIPTION
**Pull request by schmengler**
### Description

I was digging a bit for unnecessary object instantiations and one thing I found was the good old `_init` method in collections where the class name of the resource model is passed. resource models are usually shared instances, but collections misuse a "Magento\Framework\Validator\UniversalFactory" (which has a TODO comment that it should be eliminated) to create a new instance every time.

As discussed with @antonkril I introduced a `ResourceModelPool` as replacement.

For BC, the universalFactory argument stays in place, and the resourceModelPool argument has been added at the end with `null` default. The protected _universalFactory property is not used by the core anymore and is now marked as `@deprecated`

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Review by antonkril — APPROVED** [review]

File: lib/internal/Magento/Framework/Model/ResourceModel/ResourceModelPoolInterface.php [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+ */
+interface ResourceModelPoolInterface
+{
+    /**
+     * Return instance for given class name from pool
+     *
+     * @param string $resourceClassName
+     * @return AbstractResource
+     */
+    public function get($resourceClassName): AbstractResource;
```
Comment (buskamuza) [inline lib/internal/Magento/Framework/Model/ResourceModel/ResourceModelPoolInterface.php]: Please add `string` type hinting, as 2.3 supports PHP 7.0+

**Review by buskamuza — CHANGES_REQUESTED** [review]

**Comment by schmengler** [comment]
@buskamuza right, I updated the PR!

**Comment by sivaschenko** [comment]
@schmengler sorry for slow response, can you please update your branch and resolve the conflicts

**Comment by schmengler** [comment]
Will do

**Comment by sivaschenko** [comment]
@schmengler see my messages in slack as well, there is a rebased branch with this pull request that you can use https://github.com/sivaschenko/magento2/tree/resource-model-pool-rebased (but I didn't want to force push to your branch)

**Comment by schmengler** [comment]
@sivaschenko So I did it now :)

**Comment by sivaschenko** [comment]
Hi @schmengler running `bin/magento setup:di:compile` results in the following error:

    Invalid ExtensionInterface for nonexistent class Interface in file app/code/Magento/TestModuleExtensionAttributes/Api/Data/FakeExtensibleTwoInterface.php

File: lib/internal/Magento/Framework/Model/ResourceModel/ResourceModelPool.php [hunk 2]
```diff
@@ -0,0 +1,36 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+
+namespace Magento\Framework\Model\ResourceModel;
```
Comment (ihor-sviziev) [inline lib/internal/Magento/Framework/Model/ResourceModel/ResourceModelPool.php:7]: Shouldn't we have strict_types=1 there?

**Pull request merged**